### PR TITLE
`devMode` DSL extension

### PR DESCRIPTION
**Pull request by sebersole**
DSL extension for configuring dev-mode (poc)

### CODE REVIEW

**Comment by sebersole** [comment]
Sorry, should have been draft. More meant for discussion.

**Comment by sebersole** [comment]
But yes, I will squash them as well :)

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/dsl/DevMode.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+import io.quarkus.gradle.extension.QuarkusPluginExtension;
+
+/**
+ * DSL extension for configuring Quarkus dev-mode
+ *
+ * @author Steve Ebersole
+ */
+public class DevMode {
+    private final DirectoryProperty workingDirectory;
```
Comment (glefloch) [inline devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/dsl/DevMode.java:18]: Should we add a setter to make it configurable with the kotlin DSL ?

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java [hunk 2]
```diff
@@ -125,55 +122,41 @@ public QuarkusDev(
...
     public Property<File> getWorkingDirectory() {
         return workingDirectory;
     }
 
     /**
-     * @deprecated See {@link #workingDirectory}
+     * @see #workingDirectory
      */
-    @Deprecated
     public void setWorkingDir(String workingDir) {
```
Comment (glefloch) [inline devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java:144]: shouldn't we keep it as deprecated?

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java [hunk 3]
```diff
@@ -196,7 +179,7 @@ public void setPreventnoverify(boolean preventNoVerify) {
         getPreventNoVerify().set(preventNoVerify);
     }
 
-    @Input
+    @Internal
     public ListProperty<String> getJvmArguments() {
```
Comment (glefloch) [inline devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java:183]: What is the difference between `@Input` and `@Internal` for options ? shouldn't we keep a `@Input` here?

**Review by glefloch — COMMENTED** [review]
I left few comments but this looks good. The main idea is to move all configuration from task to extension right?

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java (hunk 2 above)
Comment (sebersole) [reply devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java:144]: Up to y'all.  I added the deprecation in the first place.  Depends how you want upgrading users to experience this.  

Again, I'll do whatever you want here :)

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java (hunk 3 above)
Comment (sebersole) [reply devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java:183]: Inherently Grade considers all getters as inputs of some sort, and gives you errors if they are not annotated appropriately.

`@Input` (or `@InputFile`, ...) define properties as part of the task's `Task#inputs` which are used for up-to-date checking.  

We never want these tasks to be considered up-to-date - they should always run.

`@Internal` is a way to mark a getter to tell Gradle to ignore it for up-to-date checking.  Gradle also recently added this new `@UntrackedTask` annotation which serves the same role, but at the task/class level.  To be honest, I am not sure what happens though if you use `@UntrackedTask` and leave off the specific getter annotations.  I can try if you like.

But regardless, `@Input` is not appropriate here imo

**Comment by sebersole** [comment]
> I left few comments but this looks good. The main idea is to move all configuration from task to extension right?

More or less, yes that is correct.  The hidden gain (though you'll have to tell me if this is correct and appropriate) is to only configure this once instead of 3 times (QuarkusDev, QuarkusTest and QuarkusRemoteDev).

I'd also say (IMHO anyway) using DSL over direct task configuration  is more natural for Gradle users.

**Pull request closed without merging**